client/gutenberg/editor/async-loader.jsx

```diff
@@ -0,0 +1,60 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { Component } from 'react';
+
+export const asyncLoader = ( { promises, loading, success, failure } ) =>
+	class AsyncLoader extends Component {
+		state = {
+			results: Object.keys( promises ).reduce(
+				( output, key ) => ( { ...output, [ key ]: null } ),
+				{}
+			),
+		};
+
+		componentDidMount() {
+			// here we want to make sure all promises "resolve" even when
+			// they reject because Promise.all will return a single value
+			// on the first rejection instead of an array as when they all
+			// resolve. this is for our own accounting and ensures that
+			// we wait for all promises to fulfill
+			const runners = Object.keys( promises ).map( key =>
+				promises[ key ]
+					.then( a => {
+						this.setState( state => ( { results: { ...( state || {} ).results, [ key ]: a } } ) );
```

Comment:
We should probably track if the component has unmounted and prevent the call to this.setState, or wrap all of the promises in cancelable promises and cancel them on unmount. 

Since we're currently reloading the page when entering Gutenberg, the bug isn't manifesting, but once we're able to load without a page refresh, hitting back would unmount the loader.

Reply:
Good point: 315b031f35

> or wrap all of the promises in cancelable promises and cancel them on unmount

Curious --  how do you make a cancelable promise? 3rd party lib?

Reply:
Ah, nevermind, even https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html has something on that.

Reply:
I'll stick with `_isMounted` for now tho :grin: